contracts/contracts/vault/OETHVaultCore.sol

```diff
@@ -1,26 +1,148 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.0;
 
+import { StableMath } from "../utils/StableMath.sol";
 import { VaultCore } from "./VaultCore.sol";
 
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IStrategy } from "../interfaces/IStrategy.sol";
+
 /**
  * @title OETH VaultCore Contract
  * @author Origin Protocol Inc
  */
 contract OETHVaultCore is VaultCore {
+    using SafeERC20 for IERC20;
+    using StableMath for uint256;
+
     address public immutable weth;
+    uint256 public wethAssetIndex;
 
     constructor(address _weth) {
         weth = _weth;
     }
 
+    /**
+     * @dev Caches WETH's index in `allAssets` variable.
+     *      Reduces gas usage by redeem by caching that.
+     */
+    function cacheWETHAssetIndex() external onlyGovernor {
```

Comment:
If we adding `wethAssetIndex` for gas improvements, it'd be much better if `wethAssetIndex` was an immutable. The problem with that, though, is this the immutable has to be set in the constructor where the storage slots from the proxy contract are unavailable. `wethAssetIndex` would have to passed in as a constructor param and then checked in an `initialize` or `upgrade` function that is called as part of the proxy upgrade.

Reply:
I initially thought of making it an immutable, yeah. However, we have #2012 PR that's gonna go live in the near future. When we remove other assets from the Vault, it can change the asset index. I assumed we do not want to redeploy the Vault everytime we remove an asset

Reply:
Good point, we don't want to redeploy the contract every time we remove an asset